File: frontend/src/types/ict.ts
export interface SwingPoint {
  time: string;
  index: number;
  price: number;
  type: string;
}

export interface StructureEvent {
  time: string;
  index: number;
  direction: string;
  broken_level: number;
  confidence: number;
}

export interface LiquiditySweep {
  time: string;
  index: number;
  direction: string;
  swept_level: number;
  candle_high: number;
  candle_low: number;
}

export interface FVG {
  time: string;
  index: number;
  direction: string;
  upper: number;
  lower: number;
  midpoint: number;
  is_filled?: boolean;
}

export interface IFVG {
  time: string;
  index: number;
  original_fvg_index: number;
  direction: string;
  upper: number;
  lower: number;
  midpoint: number;
}

export interface OrderBlock {
  time: string;
  index: number;
  direction: string;
  high: number;
  low: number;
  midpoint: number;
  bos_index: number;
}

export interface OTEZone {
  direction: string;
  ote_low: number;
  ote_high: number;
  fib_62: number;
  fib_705: number;
  fib_79: number;
}

export interface TradeModel {
  name: string;
  direction: string;
  confidence: number;
  reason: string;
  entry_zone?: Record<string, number> | null;
  invalid_level?: number | null;
  target_level?: number | null;
}

export interface ICTAnalysis {
  symbol: string;
  timeframe: string;
  bias: string;
  score: number;
  setup_type: string;
  swing_points: SwingPoint[];
  bos_events: StructureEvent[];
  mss_events: StructureEvent[];
  liquidity_sweeps: LiquiditySweep[];
  fvgs: FVG[];
  ifvgs: IFVG[];
  order_blocks: OrderBlock[];
  ote_zones: OTEZone[];
  session_ranges: unknown[];
  trade_models: TradeModel[];
  entry_zone: Record<string, number> | null;
  stop_loss: number | null;
  take_profit: number | null;
  explanation: string[];
}

File: backend/app/schemas/ict.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class SwingPoint(BaseModel):
    time: datetime
    index: int = Field(ge=0)
    price: float
    type: str


class StructureEvent(BaseModel):
    time: datetime
    index: int = Field(ge=0)
    direction: str
    broken_level: float
    confidence: float = Field(ge=0, le=100)


class LiquiditySweep(BaseModel):
    time: datetime
    index: int = Field(ge=0)
    direction: str
    swept_level: float
    candle_high: float
    candle_low: float


class FVG(BaseModel):
    time: datetime
    index: int = Field(ge=0)
    direction: str
    upper: float
    lower: float
    midpoint: float
    is_filled: bool = False


class IFVG(BaseModel):
    time: datetime
    index: int = Field(ge=0)
    original_fvg_index: int = Field(ge=0)
    direction: str
    upper: float
    lower: float
    midpoint: float


class OrderBlock(BaseModel):
    time: datetime
    index: int = Field(ge=0)
    direction: str
    high: float
    low: float
    midpoint: float
    bos_index: int = Field(ge=0)


class OTEZone(BaseModel):
    direction: str
    ote_low: float
    ote_high: float
    fib_62: float
    fib_705: float
    fib_79: float


class SessionRange(BaseModel):
    session_name: str
    start_time: datetime
    end_time: datetime
    high: float
    low: float
    midpoint: float


class ICTAnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    symbol: str
    timeframe: str
    bias: str
    score: int = Field(ge=0, le=100)
    setup_type: str
    swing_points: list[SwingPoint]
    bos_events: list[StructureEvent]
    mss_events: list[StructureEvent]
    liquidity_sweeps: list[LiquiditySweep]
    fvgs: list[FVG]
    ifvgs: list[IFVG]
    order_blocks: list[OrderBlock]
    ote_zones: list[OTEZone]
    session_ranges: list[SessionRange]
    entry_zone: dict[str, float] | None = None
    stop_loss: float | None = None
    take_profit: float | None = None
    explanation: list[str]
